x-pack/legacy/plugins/maps/public/layers/styles/vector/components/color/dynamic_color_form.js

```diff
@@ -13,140 +13,96 @@ import { CATEGORICAL_DATA_TYPES, COLOR_MAP_TYPE } from '../../../../../../common
+    const newColorOptions = {
+      ...styleOptions,
+      type,
+    };
+    if (type === COLOR_MAP_TYPE.ORDINAL) {
+      newColorOptions.useCustomColorRamp = useCustomColorMap;
+      newColorOptions.customColorRamp = customColorMap;
+      newColorOptions.color = color;
+    } else {
+      newColorOptions.useCustomColorPalette = useCustomColorMap;
+      newColorOptions.customColorPalette = customColorMap;
+      newColorOptions.colorCategory = color;
     }
-  }
 
-  _getColorSelector() {
-    const { onDynamicStyleChange, styleProperty } = this.props;
-    const styleOptions = styleProperty.getOptions();
+    onDynamicStyleChange(styleProperty.getStyleName(), newColorOptions);
+  };
 
+  const onFieldChange = async ({ field }) => {
+    const { name, origin, type } = field;
+    onDynamicStyleChange(styleProperty.getStyleName(), {
+      ...styleOptions,
+      field: { name, origin },
+      type: CATEGORICAL_DATA_TYPES.includes(type)
+        ? COLOR_MAP_TYPE.CATEGORICAL
+        : COLOR_MAP_TYPE.ORDINAL,
+    });
+  };
+
+  const renderColorMapSelect = () => {
     if (!styleOptions.field || !styleOptions.field.name) {
-      return;
+      return null;
     }
 
-    let colorSelect;
-    const onColorChange = colorOptions => {
-      const newColorOptions = {
-        type: colorOptions.type,
-      };
-      if (colorOptions.type === COLOR_MAP_TYPE.ORDINAL) {
-        newColorOptions.useCustomColorRamp = colorOptions.useCustomColorMap;
-        newColorOptions.customColorRamp = colorOptions.customColorMap;
-        newColorOptions.color = colorOptions.color;
-      } else {
-        newColorOptions.useCustomColorPalette = colorOptions.useCustomColorMap;
-        newColorOptions.customColorPalette = colorOptions.customColorMap;
-        newColorOptions.colorCategory = colorOptions.color;
-      }
-
-      onDynamicStyleChange(styleProperty.getStyleName(), {
-        ...styleOptions,
-        ...newColorOptions,
-      });
-    };
-
-    if (this.state.colorMapType === COLOR_MAP_TYPE.ORDINAL) {
-      const customOptionLabel = i18n.translate('xpack.maps.style.customColorRampLabel', {
-        defaultMessage: 'Custom color ramp',
-      });
-      colorSelect = (
+    if (styleOptions.type === COLOR_MAP_TYPE.ORDINAL) {
+      return (
         <ColorMapSelect
           colorMapOptions={COLOR_GRADIENTS}
-          customOptionLabel={customOptionLabel}
-          onChange={options => onColorChange(options)}
+          customOptionLabel={i18n.translate('xpack.maps.style.customColorRampLabel', {
+            defaultMessage: 'Custom color ramp',
+          })}
+          onChange={onColorMapSelect}
           colorMapType={COLOR_MAP_TYPE.ORDINAL}
           color={styleOptions.color}
           customColorMap={styleOptions.customColorRamp}
           useCustomColorMap={_.get(styleOptions, 'useCustomColorRamp', false)}
           compressed
         />
       );
-    } else if (this.state.colorMapType === COLOR_MAP_TYPE.CATEGORICAL) {
```

Comment:
by removing these else-ifs, reading the code, you really loose coherence and advantage of consistent indentation. these (categoical/ordinal) are equally weighted code-paths. putting `returns` at different levels removes all that context. removing the categorical check requires users to keep in their head the "else" iso just being able to read it.